Docs: Document which data sources work with caching

mjseaman opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/main/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the main branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

-->

**What this PR does / why we need it**:
Adds some extra docs to clarify which built-in and Enterprise data sources work with query caching, and how to tell if other third-party data sources work with query caching.

**Which issue(s) this PR fixes**:
None- this fixes an issue encountered by a user: https://raintank-corp.slack.com/archives/CNS0TK5BN/p1625613824197500

<!--

- Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

Fixes #

**Special notes for your reviewer**:

mjseaman committed:
Document which data sources work with caching

mjseaman committed:
add Prometheus and Loki to exclusions

kminehart commented on docs/sources/enterprise/query-caching.md:
```diff
@@ -27,6 +27,13 @@ You can make a panel retrieve fresh data more frequently by increasing the **Max
 - Reduced likelihood that APIs will rate-limit or throttle requests.
 
+## Data sources that work with query caching
+Query caching works for all [Enterprise data sources](https://grafana.com/grafana/plugins/?type=datasource&enterprise=1), and it works for all [built-in data sources]({{< relref "../datasources/_index.md" >}}) except for Elasticsearch, Prometheus, and Loki, which cache queries themselves.
```
We probably need to verify this: `it works for all [built-in data sources]({{< relref "../datasources/_index.md" >}}) except for Elasticsearch, Prometheus, and Loki, which cache queries themselves`.

mjseaman replied:
Went through and confirmed which built-in data sources display the Cache tab and which ones don't.

mjseaman committed:
update list of built-in data sources

I checked in Grafana v8.0.4 and these were the built-in data sources that do work with query caching.

oddlittlebird commented on docs/sources/enterprise/query-caching.md:
```diff
@@ -27,6 +27,22 @@ You can make a panel retrieve fresh data more frequently by increasing the **Max
 - Reduced API costs.
 - Reduced likelihood that APIs will rate-limit or throttle requests.
 
+## Data sources that work with query caching
```
```suggestion
## Data sources that work with query caching

```

oddlittlebird commented on docs/sources/enterprise/query-caching.md:
```diff
@@ -27,6 +27,22 @@ You can make a panel retrieve fresh data more frequently by increasing the **Max
 
+## Data sources that work with query caching
+Query caching works for all [Enterprise data sources](https://grafana.com/grafana/plugins/?type=datasource&enterprise=1), and it works for the following [built-in data sources]({{< relref "../datasources/_index.md" >}}):
+- InfluxDB
```
Alphabetize this list, please.

mjseaman replied:
Done :)

oddlittlebird commented on docs/sources/enterprise/query-caching.md:
```diff
@@ -27,6 +27,22 @@ You can make a panel retrieve fresh data more frequently by increasing the **Max
+- Google Cloud Monitoring
+- CloudWatch
+
+Some data sources,like  Elasticsearch, Prometheus, and Loki, cache queries themselves and therefore Grafana query caching does not improve performance.
```
```suggestion
Some data sources, such as Elasticsearch, Prometheus, and Loki, cache queries themselves, so Grafana query caching does not improve performance.
```

oddlittlebird commented on docs/sources/enterprise/query-caching.md:
```diff
@@ -27,6 +27,22 @@ You can make a panel retrieve fresh data more frequently by increasing the **Max
+
+Query caching also works for all data sources that include a backend. More specifically, caching works with data sources that extend the `DataSourceWithBackend` class in the plugins SDK.
+
+To tell if a data source works with query caching, follow the instructions below to **Enable and Configure query caching**. If Caching is enabled in Grafana but the Caching tab is not visible for the given data source, then query caching is not available for that data source.
```
```suggestion
To tell if a data source works with query caching, follow the instructions below to **Enable and Configure query caching**. If caching is enabled in Grafana but the Caching tab is not visible for the given data source, then query caching is not available for that data source.
```

mjseaman committed:
Apply suggestions from code review

Co-authored-by: Diana Payton <52059945+oddlittlebird@users.noreply.github.com>

mjseaman committed:
alphabetize list

oddlittlebird reviewed (approved):
Looks good!
